### ISSUE TITLE
feat: unstable_httpSubscriptionLink with headers polyfill does not update headers when reestablishing a lost connection

### ISSUE THREAD

**Issue by Nick-Lucas** [trpc/trpc#6046]
### Provide environment information

n/a

### Describe the bug

This is a known issue with event-source-polyfill https://github.com/Yaffle/EventSource/issues/137 but something we can probably fix in tRPC-land

Essentially EventSource is given headers and those headers are passed at instantiation. When a connection is lost EventSource will opaquely recover and reconnect, however it will use the original headers which may be outdated. This is a problem when the Authorization header contains a token which has expired since the EventSource was created

### Link to reproduction

https://github.com/trpc/trpc/pull/6047

### To reproduce

Pass a short lived header to unstable_httpSubscriptionLink.eventSourceOptions when using the polyfill, and then reboot the backend server. The header will be re-sent when connection is reestablished

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/6046">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/6046/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/6046/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
Auth through headers should still work.... but I see this is a potentially issue when using polyfill

Maybe we could have an option to recreate the `EventSource` on any non-4xx errors (where the callbacks are called again) and gracefully resume the stream

**Comment by Nick-Lucas** [comment]
> Auth through headers should still work.... but I see this is a potentially issue when using polyfill
> 
> Maybe we could have an option to recreate the `EventSource` on any non-4xx errors (where the callbacks are called again) and gracefully resume the stream

PR now has a first approach to restarting for 401/403 errors, if I can get this working then we can look at how best to expose control too

Maybe like this:

```ts
unstable_httpSubscriptionLink({
  url: opts.httpUrl,
  transformer: superjson,
  eventSourceOptions() {
    return {
      headers: {
        // etc
      },
    } as EventSourcePolyfillInit;
  },
  shouldRecreate(event) {
    return [401, 403].includes(event.status)
  }
})
```

**Comment by KATT** [comment]
That api LGTM

**Comment by KATT** [comment]
- `shouldRecreate()` should probably accept a `boolean | Promise<boolean>`
- We should avoid calling `shouldRecreate()` twice in parallel - can use a deferred or something
